File: 06 - Adapter Pattern.ts
// Target interface
interface MediaPlayer {
    play(filename: string): void;
}

// Existing MP3 player
class MP3Player implements MediaPlayer {
    public play(filename: string): void {
        console.log(`Playing MP3 file: ${filename}`);
    }
}

// Modern player (Adaptee)
class ModernPlayer {
    public playMP4(filename: string): void {
        console.log(`Playing MP4 file: ${filename}`);
    }
}

// Adapter
class ModernPlayerAdapter implements MediaPlayer {
    private modernPlayer: ModernPlayer;
    
    constructor(player: ModernPlayer) {
        this.modernPlayer = player;
    }
    
    public play(filename: string): void {
        if (filename.endsWith('.mp4')) {
            this.modernPlayer.playMP4(filename);
        } else {
            console.log('Invalid media format');
        }
    }
}

// Client code WITH pattern
class ClientWithAdapter {
    public static main(): void {
        const player: MediaPlayer = new MP3Player();
        const modernPlayerAdapter: MediaPlayer = new ModernPlayerAdapter(new ModernPlayer());
        
        // Unified interface for all players
        player.play('song.mp3');
        modernPlayerAdapter.play('video.mp4');
        
        // Can handle any player through the same interface
        ClientWithAdapter.playMedia('song.mp3', player);
        ClientWithAdapter.playMedia('video.mp4', modernPlayerAdapter);
    }
    
    // Can write methods that work with any player
    public static playMedia(filename: string, player: MediaPlayer): void {
        player.play(filename);
    }
}

// Execute the main method
ClientWithAdapter.main();

File: 06 - Adapter Pattern.java
// Target interface
interface MediaPlayer {
    void play(String filename);
}

// Existing MP3 player
class MP3Player implements MediaPlayer {
    public void play(String filename) {
        System.out.println("Playing MP3 file: " + filename);
    }
}

// Modern player (Adaptee)
class ModernPlayer {
    public void playMP4(String filename) {
        System.out.println("Playing MP4 file: " + filename);
    }
}

// Adapter
class ModernPlayerAdapter implements MediaPlayer {
    private ModernPlayer modernPlayer;
    
    public ModernPlayerAdapter(ModernPlayer player) {
        this.modernPlayer = player;
    }
    
    @Override
    public void play(String filename) {
        if (filename.endsWith(".mp4")) {
            modernPlayer.playMP4(filename);
        } else {
            System.out.println("Invalid media format");
        }
    }
}

class Client {
    public static void main(String[] args) {
        MediaPlayer player = new MP3Player();
        MediaPlayer modernPlayerAdapter = 
            new ModernPlayerAdapter(
                new ModernPlayer());
        
        // Unified interface for all players
        player.play("song.mp3");
        modernPlayerAdapter.play("video.mp4");
        
        // Can handle any player through the same interface
        playMedia("song.mp3", player);
        playMedia("video.mp4", modernPlayerAdapter);
    }
    
    public static void playMedia(
        String filename, MediaPlayer player) {
        player.play(filename);
    }
}